### DISCUSSION TITLE
Unimplemented api features

### DISCUSSION THREAD

**Discussion by coeur-de-loup** [leolabs/ableton-js#77] [Q&A] [answered]
Hi Leo!
I've been using your script with a web app regularly on stage for about 2 months now. It's absolutely fantastic. I've said it in my last message, but thanks again for the hard work!

I'm wondering if there is a way to call the Ableton api on methods and properties that are not yet implemented in the script namespace? 

For example, I managed to receive a Browser object, but its a string that looks like a python memory address, not the actual Browser object.

Also, I'd like to observe and set tracks mixer_device props, like volume. Is there a way of doing this?

**Comment by leolabs** [comment, marked as answer]
Thank you for your feedback, I'm happy to hear that you find this library useful!

Generally, all properties that contain strings, booleans, and numbers, and all functions in the supported namespaces can be accessed without making any change to this library. However, if the property is an object like the browser you're trying to access, the library needs to know how to serialize the object.

If there's no defined serialization, ableton-js will try to convert the property into a string which is what you got when you tried to access the browser property.

I should be able to make a PR for this soon, but going through the [API specification of Live](https://nsuspray.github.io/Live_API_Doc/11.0.0.xml#Live-MixerDevice-MixerDevice), it seems like the `volume` property can't be observed unfortunately. Only `crossfade_assign`, `panning_mode`, and `sends` can be observed.

I hope this helped. Let me know if you have any further questions!

**Comment by coeur-de-loup** [comment]
Hi,

Thanks for the mixer_device!

I was curious about the volume not being observable, control surfaces like the push and the Novation slMk3 actually listens and properly shows volume changes. You're right, it is not in the documentation... But looking in the remote scripts, some things that looks like they are related to observing the volume. Is this something that could be implemented you think? I'm just starting python, so I don't think I have the qualification to submit PRs just yet (but can't wait to try to help if it is permitted!). Maybe I ask to much, but I'm really curious about the possibility of such  use of undocumented features!


```Python
from ableton.v2.base import clamp, listens, liveobj_valid, old_round
import ableton.v2.control_surface.components as ChannelStripComponentBase 

 @listens('value')
    def __on_volume_changed(self):
        self._update_volume_led()
        
self._ChannelStripComponent__on_volume_changed.subject = track.mixer_device.volume if liveobj_valid(track) else None
self._ChannelStripComponent__on_volume_changed()
```

On another note, If you find features or issues that could be coded by beginner/intermediate programmer, I would gladly help! I just started in CS at university and this project would a the perfect learning experience for me!

Thanks you!

**Reply by leolabs** [reply]
That's interesting! I could imagine that the `@listens` decorator checks the value on a regular interval and runs the callback function when it notices that it has changed. We could do something similar with ableton-js, but I'm not sure about the performance implications yet.

> If you find features or issues that could be coded by beginner/intermediate programmer, I would gladly help! I just started in CS at university and this project would a the perfect learning experience for me!

There are still classes in the unofficial docs that ableton-js doesn't support yet, so if you find any that you'd like to try implementing support for, you could give it a try. The structure is fairly simple, you can use the other classes as examples for how it works. And if you have any questions, feel free to ask me or submit a draft PR.

**Reply by leolabs** [reply]
I just saw in the Max for Live docs that the mixer device params are actually a type of DeviceParameter which would offer a lot more flexibility when it comes to observing and setting values:

![CleanShot 2023-01-12 at 00 13 07](https://user-images.githubusercontent.com/541628/211938479-d1e510e9-e28c-48e5-87b6-53ff8713025b.png)

I'll play around with this later, but this could be a good solution.

**Reply by leolabs** [reply]
It worked! I've released a new version that adds support for this. You can now observe any mixer value like this:

```ts
const tracks = await ableton.song.get("tracks");
const mixer = await tracks[0].get("mixer_device");
const volume = await mixer.get("volume");

volume.addListener("value", (v) => console.log(v));
```
